indispensable/src/commonMain/kotlin/at/asitplus/signum/indispensable/asn1/PemEncodable.kt

```diff
@@ -0,0 +1,110 @@
+package at.asitplus.signum.indispensable.asn1
+
+import at.asitplus.KmmResult
+import at.asitplus.catching
+import kotlin.io.encoding.Base64
+import kotlin.io.encoding.ExperimentalEncodingApi
+
+/**
+ * Specialization of [Asn1Encodable], able to produce PEM-encoded strings
+ * as per [RFC 1421](https://datatracker.ietf.org/doc/html/rfc1421#section-4.4).
+ * Use in tandem with [PemDecodable].
+ */
+interface PemEncodable<A : Asn1Element> : Asn1Encodable<A> {
+
+    /**
+     * Encapsulation boundary string.
+     * Will be automatically fenced.
+     * Make sure it matches the corresponding [PemDecodable.ebString]!
+     */
+    val ebString: String
+
+    /**
+     * To be implemented for custom encoding of PEM-encoded payload.
+     * Defaults to [encodeToDer]
+     */
+    @Throws(Throwable::class)
+    fun binaryEncodePayload(): ByteArray = encodeToDer()
+}
+
+/**
+ * Specialization of [Asn1Decodable], able to parse PEM-encoded strings
+ * as per [RFC 1421](https://datatracker.ietf.org/doc/html/rfc1421).
+ * Use in tandem with [PemEncodable].
+ */
+interface PemDecodable<A : Asn1Element, T : PemEncodable<A>> : Asn1Decodable<A, T> {
+
+    /**
+     * Encapsulation boundary string.
+     * Will be automatically fenced.
+     * Make sure it matches the corresponding [PemEncodable.ebString]!
+     */
+    val ebString: String
+
+    /**
+     * To be implemented for custom decoding of PEM-encoded payload.
+     * Defaults to [decodeFromDer] and validates pre- and postEb
+     */
+    @Throws(Throwable::class)
+    fun binaryDecodePayload(ebString: String, src: ByteArray): T {
+        require(ebString == this.ebString) { "PEM encoded EB string does not match with ${this.ebString}: $ebString" }
+        return decodeFromDer(src)
+    }
+}
+
+
+private const val FENCE_PRE = "-----BEGIN "
+private const val FENCE_POST = "-----END "
+private const val FENCE_AFTER = "-----"
+
+/**-----BEGIN [PemEncodable.ebString]-----*/
+val PemEncodable<*>.preEB: String get() = "$FENCE_PRE$ebString$FENCE_AFTER"
+
+/**-----END [PemEncodable.ebString]-----*/
+val PemEncodable<*>.postEB: String get() = "$FENCE_POST$ebString$FENCE_AFTER"
+
+/**-----BEGIN [PemDecodable.ebString]-----*/
+val PemDecodable<*, *>.preEB: String get() = "$FENCE_PRE$ebString$FENCE_AFTER"
+
+/**-----END [PemDecodable.ebString]-----*/
+val PemDecodable<*, *>.postEB: String get() = "$FENCE_POST$ebString$FENCE_AFTER"
+
+/**
+ * Encodes this [PemEncodable] into a PEM-encoded string
+ */
+@OptIn(ExperimentalEncodingApi::class)
+fun PemEncodable<*>.encodeToPEM(): KmmResult<String> = catching {
+    "$preEB\n" + Base64.encode(binaryEncodePayload()).chunked(64)
+        .joinToString(separator = "\n", postfix = "\n") + postEB
+}
+
+/**
+ * Reads the first line of the passed string and tries to extract the encapsulation boundary string
+ */
+@Throws(Throwable::class)
+private fun PemDecodable<*, *>.peekEbString(src: String): String {
```

Comment:
why is this not part of `decodeFromPem`? we are splitting by lines a total of three times

Reply:
so it can't be overridden